[Writing Tools] There is a brief flicker where there is no text during a rewrite animation

rr-codes opened the pull request:
#### e1f85b19921afbf117601385f2011c05e537ca78
<pre>
[Writing Tools] There is a brief flicker where there is no text during a rewrite animation
<a href="https://bugs.webkit.org/show_bug.cgi?id=280038">https://bugs.webkit.org/show_bug.cgi?id=280038</a>
<a href="https://rdar.apple.com/136334757">rdar://136334757</a>

Reviewed by Aditya Keerthi, Wenson Hsieh, and Abrar Rahman Protyasha.

A sample sequence of events with regards to when intelligence text animations begin and end is

* initial animation created
* initial animation removed [0 ongoing animations]
* source animation created
* final animation created
* final animation removed
* initial animation created
* source animation removed
* initial animation removed [0 ongoing animations]
* source animation created
* final animation created
* final animation removed
* initial animation created
* source animation removed
* initial animation removed [0 ongoing animations]
* source animation created
* final animation created
* final animation removed
* source animation removed [0 ongoing animations]

Prior to 283274@main, the text effect view was never being removed. Following that commit, it is removed
whenever an animation was removed such that there are 0 ongoing animations. As a result, after each
initial animation removal, the effect view is removed momentarily, and then is re-added when the following
source animation gets added. This manifests visually as a flicker of the text content in the message.

Fix by only removing the effect view when there are 0 ongoing animations following a source or destination
animation and not following an initial animation.

* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
(-[WKTextAnimationManager removeTextAnimationForAnimationID:]):
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
(-[WKTextAnimationManager textPreviewForRect:completion:]):
(-[WKTextAnimationManager updateIsTextVisible:forChunk:completion:]):

Canonical link: <a href="https://commits.webkit.org/284003@main">https://commits.webkit.org/284003@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0f6041dff94681202e1bc3b3acc91affb0a755fb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67964 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47356 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20623 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72023 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19109 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55152 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18925 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54326 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12746 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71031 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43375 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34790 "Found 1 new API test failure: /WPE/TestResources:/webkit/WebKitWebView/sync-request-on-max-conns (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40043 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17466 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62018 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16492 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73719 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11935 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15780 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61788 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11973 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58810 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61800 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9693 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10358 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43156 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44230 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45425 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43972 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
Thoughts on making this a little less repetitive?

```
- (void)removeTextAnimationForAnimationID:(NSUUID *)uuid andHideTextAnimationViewIfNeeded:(BOOL)hideTextAnimationViewIfNeeded
{
    RetainPtr effectData = [_chunkToEffect objectForKey:uuid];
    if (effectData) {
        [_effectView removeEffect:[effectData effectID]];
        [_chunkToEffect removeObjectForKey:uuid];
    }

    if (hideTextAnimationViewIfNeeded == YES && ![self hasActiveTextAnimationType])
       [self hideTextAnimationView];
}

- (void)removeTextAnimationForAnimationID:(NSUUID *)uuid
{
    [self removeTextAnimationForAnimationID:uuid andHideTextAnimationViewIfNeeded:NO];
}
```

rr-codes committed:
[Writing Tools] There is a brief flicker where there is no text during a rewrite animation
https://bugs.webkit.org/show_bug.cgi?id=280038
rdar://136334757

Reviewed by Aditya Keerthi, Wenson Hsieh, and Abrar Rahman Protyasha.

A sample sequence of events with regards to when intelligence text animations begin and end is

* initial animation created
* initial animation removed [0 ongoing animations]
* source animation created
* final animation created
* final animation removed
* initial animation created
* source animation removed
* initial animation removed [0 ongoing animations]
* source animation created
* final animation created
* final animation removed
* initial animation created
* source animation removed
* initial animation removed [0 ongoing animations]
* source animation created
* final animation created
* final animation removed
* source animation removed [0 ongoing animations]

Prior to 283274@main, the text effect view was never being removed. Following that commit, it is removed
whenever an animation was removed such that there are 0 ongoing animations. As a result, after each
initial animation removal, the effect view is removed momentarily, and then is re-added when the following
source animation gets added. This manifests visually as a flicker of the text content in the message.

Fix by only removing the effect view when there are 0 ongoing animations following a source or destination
animation and not following an initial animation.

* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
(-[WKTextAnimationManager removeTextAnimationForAnimationID:]):
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
(-[WKTextAnimationManager textPreviewForRect:completion:]):
(-[WKTextAnimationManager updateIsTextVisible:forChunk:completion:]):

Canonical link: https://commits.webkit.org/284003@main
